File: Tutorials/Codes/Modern_Control_Course/Chapter25/Lesson3/Chapter25_Lesson3.py
"""
Chapter25_Lesson3.py

Trade-offs in state-feedback design:
speed of response vs. control effort vs. sensitivity.

Dependencies:
    pip install numpy scipy matplotlib

Optional extension:
    pip install control
"""

import numpy as np
from scipy.signal import place_poles
from scipy.linalg import solve_continuous_lyapunov
import matplotlib.pyplot as plt


def closed_loop_metrics(A, B, poles, x0):
    """Return K, closed-loop eigenvalues, effort Gramian, and J_u for u=-Kx."""
    result = place_poles(A, B, poles, method="YT")
    K = result.gain_matrix
    Ac = A - B @ K

    # P_u solves Ac.T P + P Ac + K.T K = 0.
    P_u = solve_continuous_lyapunov(Ac.T, -(K.T @ K))
    effort = (x0.T @ P_u @ x0).item()

    eigvals = np.linalg.eigvals(Ac)
    speed = -np.max(np.real(eigvals))
    k_norm = float(np.linalg.norm(K, 2))

    return K, Ac, eigvals, speed, k_norm, effort, P_u


def rk4_simulate(Ac, K, x0, tf=10.0, dt=0.002):
    """Simulate x_dot = Ac x and u = -Kx."""
    n_steps = int(tf / dt) + 1
    t = np.linspace(0.0, tf, n_steps)
    x = np.zeros((n_steps, len(x0)))
    u = np.zeros(n_steps)
    x[0, :] = x0.ravel()

    def f(xv):
        return Ac @ xv

    for k in range(n_steps - 1):
        xv = x[k, :]
        k1 = f(xv)
        k2 = f(xv + 0.5 * dt * k1)
        k3 = f(xv + 0.5 * dt * k2)
        k4 = f(xv + dt * k3)
        x[k + 1, :] = xv + (dt / 6.0) * (k1 + 2*k2 + 2*k3 + k4)
        u[k] = float(-(K @ x[k, :].reshape(-1, 1))[0, 0])
    u[-1] = float(-(K @ x[-1, :].reshape(-1, 1))[0, 0])
    return t, x, u


def monte_carlo_eigenvalue_sensitivity(A, B, K, samples=200, eps=0.02, seed=7):
    """Perturb A and B and measure the spread of closed-loop eigenvalues."""
    rng = np.random.default_rng(seed)
    eigs = []
    for _ in range(samples):
        dA = eps * rng.standard_normal(A.shape)
        dB = eps * rng.standard_normal(B.shape)
        Ac_perturbed = (A + dA) - (B + dB) @ K
        eigs.append(np.linalg.eigvals(Ac_perturbed))
    return np.array(eigs)


def main():
    # Plant: mass-spring-damper type state-space model.
    A = np.array([[0.0, 1.0],
                  [-2.0, -0.4]])
    B = np.array([[0.0],
                  [1.0]])
    x0 = np.array([[1.0],
                   [0.0]])

    pole_sets = {
        "slow": np.array([-1.0 + 1.0j, -1.0 - 1.0j]),
        "medium": np.array([-3.0 + 3.0j, -3.0 - 3.0j]),
        "fast": np.array([-6.0 + 6.0j, -6.0 - 6.0j]),
    }

    print("State-feedback trade-off table")
    print("case       speed       ||K||2       J_u(x0)        poles")
    print("-" * 74)

    trajectories = {}
    for name, poles in pole_sets.items():
        K, Ac, eigvals, speed, k_norm, effort, _ = closed_loop_metrics(A, B, poles, x0)
        print(f"{name:7s} {speed:10.4f} {k_norm:12.4f} {effort:12.4f}   {eigvals}")
        t, x, u = rk4_simulate(Ac, K, x0)
        trajectories[name] = (t, x, u, K, Ac)

    plt.figure()
    for name, (t, x, u, _, _) in trajectories.items():
        plt.plot(t, x[:, 0], label=f"{name}: x1")
    plt.xlabel("time (s)")
    plt.ylabel("state x1")
    plt.title("Faster poles reduce settling time")
    plt.legend()
    plt.grid(True)

    plt.figure()
    for name, (t, x, u, _, _) in trajectories.items():
        plt.plot(t, u, label=f"{name}: u")
    plt.xlabel("time (s)")
    plt.ylabel("control input u")
    plt.title("Faster poles require larger control action")
    plt.legend()
    plt.grid(True)

    for name in ["medium", "fast"]:
        _, _, _, K, _ = trajectories[name]
        eig_cloud = monte_carlo_eigenvalue_sensitivity(A, B, K)
        spread = np.std(eig_cloud.reshape(-1))
        print(f"eigenvalue cloud std for {name:6s}: {spread:.5f}")

    plt.show()


if __name__ == "__main__":
    main()

File: Tutorials/Codes/Modern_Control_Course/Chapter25/Lesson3/Chapter25_Lesson3.m
% Chapter25_Lesson3.m
%
% Trade-offs in state-feedback design:
% speed of response vs. control effort vs. sensitivity.
%
% MATLAB requirements:
%   Control System Toolbox for place, ss, initial, lyap.
% Optional:
%   Simulink for the final programmatic model section.

clear; clc; close all;

A = [0 1; -2 -0.4];
B = [0; 1];
C = [1 0];
D = 0;
x0 = [1; 0];

poleSets = {
    'slow',   [-1+1i, -1-1i];
    'medium', [-3+3i, -3-3i];
    'fast',   [-6+6i, -6-6i]
};

fprintf('State-feedback trade-off table\n');
fprintf('%10s%12s%14s%14s%16s\n', 'case', 'speed', 'norm(K)', 'J_u(x0)', 'eig(A-BK)');

figure; hold on; grid on;
title('Faster poles reduce settling time');
xlabel('time (s)'); ylabel('state x_1');

figure; hold on; grid on;
title('Faster poles require larger input');
xlabel('time (s)'); ylabel('control input u');

for i = 1:size(poleSets, 1)
    name = poleSets{i, 1};
    poles = poleSets{i, 2};

    K = place(A, B, poles);
    Ac = A - B*K;

    % P_u solves Ac' P + P Ac + K'K = 0.
    P_u = lyap(Ac', K'*K);
    J_u = x0' * P_u * x0;

    eigVals = eig(Ac);
    speed = -max(real(eigVals));

    fprintf('%10s%12.4f%14.4f%14.4f    [%8.3f%+8.3fi, %8.3f%+8.3fi]\n', ...
        name, speed, norm(K, 2), J_u, ...
        real(eigVals(1)), imag(eigVals(1)), real(eigVals(2)), imag(eigVals(2)));

    sysCL = ss(Ac, B, C, D);
    t = linspace(0, 10, 2000);
    [y, tOut, x] = initial(sysCL, x0, t);
    u = -(K * x')';

    figure(1); plot(tOut, x(:, 1), 'DisplayName', name);
    figure(2); plot(tOut, u, 'DisplayName', name);
end

figure(1); legend('Location', 'best');
figure(2); legend('Location', 'best');

% Sensitivity experiment: perturb A and B and inspect closed-loop poles.
rng(7);
epsLevel = 0.02;
samples = 200;

for i = 2:3
    name = poleSets{i, 1};
    poles = poleSets{i, 2};
    K = place(A, B, poles);
    eigCloud = zeros(samples, 2);

    for s = 1:samples
        dA = epsLevel * randn(size(A));
        dB = epsLevel * randn(size(B));
        eigCloud(s, :) = eig((A + dA) - (B + dB)*K).';
    end

    fprintf('eigenvalue cloud std for %s design: %.5f\n', name, std(eigCloud(:)));
end

% Optional Simulink sketch:
% This creates a minimal closed-loop state-space model if Simulink is installed.
if exist('simulink', 'file') == 4
    model = 'Chapter25_Lesson3_Simulink';
    if bdIsLoaded(model)
        close_system(model, 0);
    end
    new_system(model);
    open_system(model);

    % Use the medium design for the Simulink block.
    K = place(A, B, [-3+3i, -3-3i]);
    Ac = A - B*K;

    add_block('simulink/Continuous/State-Space', [model '/ClosedLoopPlant'], ...
        'A', mat2str(Ac), 'B', mat2str(B), 'C', mat2str(C), 'D', mat2str(D), ...
        'X0', mat2str(x0), 'Position', [140 100 280 160]);

    add_block('simulink/Sources/Constant', [model '/ZeroInput'], ...
        'Value', '0', 'Position', [40 115 90 145]);

    add_block('simulink/Sinks/Scope', [model '/Scope'], ...
        'Position', [340 105 390 155]);

    add_line(model, 'ZeroInput/1', 'ClosedLoopPlant/1');
    add_line(model, 'ClosedLoopPlant/1', 'Scope/1');

    set_param(model, 'StopTime', '10');
    save_system(model);
    fprintf('Created optional Simulink model: %s.slx\n', model);
end
